apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommand.ExecutableCommand.cs

```diff
@@ -178,6 +179,38 @@ internal async Task<long> ExecutePartitionedUpdateAsync(CancellationToken cancel
                 return await transaction.ExecuteDmlAsync(request, cancellationToken, CommandTimeout).ConfigureAwait(false);
             }
 
+            internal async Task<string> GetInstanceEndpointUriAsync()
+            {
+                InstanceAdminClient instanceAdminClient =
+                    await InstanceAdminClient.CreateAsync().ConfigureAwait(false);
+                // Create request for instance with endpoints.
+                GetInstanceRequest request = new GetInstanceRequest
+                {
+                    InstanceName = new InstanceName(Connection.Project, Connection.SpannerInstance),
+                    FieldMask = new FieldMask { Paths = { "endpoint_uris" } }
+                };
+                try
+                {
+                    // Get list of available endpointUris from GetInstance.
+                    Instance instance = instanceAdminClient.GetInstance(request);
+                    return instance.EndpointUris.Count == 0 ? null : instance.EndpointUris[0];
+                }
+                catch (RpcException gRpcException)
+                {
+                    if (gRpcException.StatusCode == StatusCode.PermissionDenied)
+                    {
+#pragma warning disable CS1030 // #warning directive
+#warning PermissionDenied exception
+                        // Print warning message with PermissionDenied
+                        // status in Debug mode.
+                        Console.WriteLine(gRpcException.Message);
```

Comment:
We have a Logger. You can access it from the `SessionPoolManager` if I remember correctly, you should use that one.